python/ray/tune/trial_runner.py

```diff
@@ -227,17 +226,23 @@ def _validate_resume(self, resume_type):
 
     @classmethod
     def checkpoint_exists(cls, directory):
-        if not os.path.exists(directory):
-            return False
-        return any(
-            (fname.startswith("experiment_state") and fname.endswith(".json"))
-            for fname in os.listdir(directory))
+        return file_exists(directory, cls.CKPT_FILE_REGEX)
+
+    @classmethod
+    def searcher_checkpoint_exists(cls, directory):
+        return file_exists(directory, cls.SEARCHER_CKPT_FILE_REGEX)
+
+    def _find_newest_ckpt(cls, directory):
+        return find_newest_dated_path(directory, cls.CKPT_FILE_REGEX)
+
+    def _find_newest_searcher_ckpt(cls, directory):
+        return find_newest_dated_path(directory, cls.SEARCHER_CKPT_FILE_REGEX)
 
     def add_experiment(self, experiment):
-        if not self._resumed:
-            self._search_alg.add_configurations([experiment])
-        else:
-            logger.info("TrialRunner resumed, ignoring new add_experiment.")
+        if self._resumed:
+            experiment.spec["num_samples"] = max(
+                0, experiment.spec["num_samples"] - len(self._trials))
+        self._search_alg.add_configurations([experiment])
```

Comment:
I think this actually causes a Tune restored run that uses `grid_search` to repeat, which is undesired behavior. 

I'll clean this PR up later next week.